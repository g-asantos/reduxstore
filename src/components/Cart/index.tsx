import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FiPlus, FiMinus } from 'react-icons/fi';
import { IState } from '../../store';
import {
  addProductToCartRequest,
  removeProductFromCart,
} from '../../store/modules/cart/actions';
import { ICartItem } from '../../store/modules/cart/types';
import {
  Button,
  ButtonContainer,
  Container,
  QuantityText,
  Table,
  TableData,
  TableRow,
} from './styles';

const Cart: React.FC = () => {
  const cart = useSelector<IState, ICartItem[]>(state => state.cart.items);
  const failedStock = useSelector<IState, string[]>(
    state => state.cart.failedStockCheck,
  );
  const dispatch = useDispatch();

  const hasFailedStockCheck = useCallback(
    (productId: string) => {
      return failedStock.includes(productId);
    },
    [failedStock],
  );

  useSelector<IState, boolean>(state => {
    return state.cart.failedStockCheck.length === 0;
  });

  const handleAddProductToCart = useCallback(
    product => {
      dispatch(addProductToCartRequest(product));
    },
    [dispatch],
  );

  const handleRemoveProductFromCart = useCallback(
    product => {
      dispatch(removeProductFromCart(product));
    },
    [dispatch],
  );
  return (
    <Container>
      <Table>
        <tbody>
          {cart.map(item => {
            return (
              <TableRow key={item.product.id}>
                <td>
                  <img
                    src={item.product.image}
                    style={{ width: 100, height: 100 }}
                    alt={item.product.name}
                  />
                </td>
                <TableData>
                  {item.product.name}
                  <ButtonContainer>
                    <Button
                      itemID="add_test"
                      type="button"
                      onClick={() => handleAddProductToCart(item.product)}
                    >
                      <FiPlus size={16} />
                    </Button>
                    <QuantityText>{item.quantity}</QuantityText>
                    <Button
                      itemID="remove_test"
                      type="button"
                      onClick={() => handleRemoveProductFromCart(item.product)}
                    >
                      <FiMinus size={16} />
                    </Button>
                  </ButtonContainer>
                </TableData>
                <TableData>
                  {(Number(item.product.value) * Number(item.quantity)).toFixed(
                    2,
                  )}
                </TableData>
                {hasFailedStockCheck(item.product.id) && (
                  <td style={{ color: 'red' }}>Out of Stock</td>
                )}
              </TableRow>
            );
          })}
        </tbody>
      </Table>
    </Container>
  );
};

export default Cart;
