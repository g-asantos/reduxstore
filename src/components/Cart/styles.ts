import styled from 'styled-components';

export const Container = styled.div``;

export const Table = styled.table``;

export const TableRow = styled.tr`
  display: flex;
  margin-top: 20px;
  border-bottom: 1px solid #6c757d;
`;

export const TableData = styled.td`
  margin: 0 30px 0 30px;
`;

export const ButtonContainer = styled.div`
  display: flex;
  margin-top: 30px;
`;

export const Button = styled.button`
  border: 0;
  background-color: #f5f6f7;
`;

export const QuantityText = styled.p`
  margin: 0 15px 0 15px;
`;
