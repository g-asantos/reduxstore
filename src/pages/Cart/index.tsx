import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FiPlus, FiMinus } from 'react-icons/fi';
import { IState } from '../../store';
import {
  addProductToCart,
  removeProductFromCart,
} from '../../store/modules/cart/actions';
import { ICartItem } from '../../store/modules/cart/types';
import {
  Button,
  ButtonContainer,
  Container,
  QuantityText,
  Table,
  TableContainer,
  TableData,
  TableRow,
} from './styles';

const Cart: React.FC = () => {
  const cart = useSelector<IState, ICartItem[]>(state => state.cart.items);
  const dispatch = useDispatch();

  const handleAddProductToCart = useCallback(
    product => {
      dispatch(addProductToCart(product));
    },
    [dispatch],
  );

  const handleRemoveProductFromCart = useCallback(
    product => {
      dispatch(removeProductFromCart(product));
    },
    [dispatch],
  );

  return (
    <Container>
      <Table>
        <tbody>
          {cart.map(item => (
            <TableContainer>
              <TableRow key={item.product.id}>
                <img
                  src={item.product.image}
                  style={{ width: 100, height: 100 }}
                  alt="something"
                />
                <TableData>
                  {item.product.name}
                  <ButtonContainer>
                    <Button
                      type="button"
                      onClick={() => handleAddProductToCart(item.product)}
                    >
                      <FiPlus size={16} />
                    </Button>
                    <QuantityText>{item.quantity}</QuantityText>
                    <Button
                      type="button"
                      onClick={() => handleRemoveProductFromCart(item.product)}
                    >
                      <FiMinus size={16} />
                    </Button>
                  </ButtonContainer>
                </TableData>
                <TableData>
                  {(item.product.value * item.quantity).toFixed(2)}
                </TableData>
              </TableRow>
            </TableContainer>
          ))}
        </tbody>
      </Table>
    </Container>
  );
};

export default Cart;
