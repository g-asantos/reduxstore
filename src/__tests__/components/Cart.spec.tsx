/* eslint-disable prettier/prettier */
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Store, AnyAction } from 'redux';
import configureStore from 'redux-mock-store';
import Cart from '../../components/Cart'

const mockStore = configureStore([]);



describe('Cart', () => {
  let store: Store<any, AnyAction>;
  let component: any;

  beforeEach(() => {
    store = mockStore({
      cart: {
        items:  [{
          product:{
            id: 1,
            name: 'test',
            image: 'test',
            details: 'test',
            value: 20
                    },
          quantity: 2,
        }],
        failedStockCheck: []
      },
      },
    );

    store.dispatch = jest.fn()

    component = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Cart />
        </BrowserRouter>
      </Provider>
    );
  });

  it('should render with given state from Redux Store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should dispatch an action on add button click', () => {

    const addButton = component.root.findByProps({itemID: 'add_test'})
    renderer.act(() => {
      addButton.props.onClick();
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1)
  })

  it('should dispatch an action on remove button click', () => {

    const addButton = component.root.findByProps({itemID: 'remove_test'})
    renderer.act(() => {
      addButton.props.onClick();
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1)
  })

});
