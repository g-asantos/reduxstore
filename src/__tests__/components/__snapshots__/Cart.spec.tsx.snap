// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should render with given state from Redux Store 1`] = `
<div
  className="sc-bdnylx"
>
  <table
    className="sc-gtssRu"
  >
    <tbody>
      <tr
        className="sc-dlnjPT eMGgmT"
      >
        <td>
          <img
            alt="test"
            src="test"
            style={
              Object {
                "height": 100,
                "width": 100,
              }
            }
          />
        </td>
        <td
          className="sc-hKFyIo eRPwzg"
        >
          test
          <div
            className="sc-eCApGN keFIjH"
          >
            <button
              className="sc-jSFkmK dafcKe"
              itemID="add_test"
              onClick={[Function]}
              type="button"
            >
              <svg
                fill="none"
                height={16}
                size={16}
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width={16}
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="12"
                  x2="12"
                  y1="5"
                  y2="19"
                />
                <line
                  x1="5"
                  x2="19"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
            <p
              className="sc-gKAblj fBUFXf"
            >
              2
            </p>
            <button
              className="sc-jSFkmK dafcKe"
              itemID="remove_test"
              onClick={[Function]}
              type="button"
            >
              <svg
                fill="none"
                height={16}
                size={16}
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width={16}
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="5"
                  x2="19"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
          </div>
        </td>
        <td
          className="sc-hKFyIo eRPwzg"
        >
          40.00
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
